### DISCUSSION TITLE
Validate only if one property equals to a value

### DISCUSSION THREAD

**Discussion by CarterLi** [json-schema-org/community#417] [Q&A]
I'm creating JSON schema for [the JSON configuration file of my project](https://github.com/fastfetch-cli/fastfetch/issues/471)

This is the WIP schema: https://github.com/fastfetch-cli/fastfetch/blob/dev/src/data/json_schema.jsonc

---

The question:

My program have many modules. Each module have its own type. eg

```json
[
    {
        "type": "CPU",
        "temp": true
    },
    {
        "type": "GPU",
        "temp": true
    },
    {
        "type": "LocalIp",
        "showIpv6": true,
        "namePrefix": "en0"
    }
]
```

`"type"` specifies the module type. This is the only required property.

Only if the module type is either `"CPU"` or `"GPU"`, a user can specify another property `"temp"`. Only if the module type is `"LocalIp"`, a user can specify another two properties `"showIpv6"` and `"namePrefix"`. Module `CPU` and `GPU` cannot have property `showIpv6` and `namePrefix`. Module `LocalIp` cannot have property `temp`.

What's the best way to do that?

**Comment by jdesrosiers** [comment]
Since this is for a config file, I assume the intended validator is your editor/IDE of choice. Unfortunately, a lot of editors/IDEs haven't implemented some of the newer JSON Schema features (specifically, `unevaluatedProperties`) that would be useful in cases like this. Therefore, I'd suggest something like the following.

```json
{
  "anyOf": [
    {
      "type": "object",
      "properties": {
        "type": { "enum": ["CPU", "GPU"] },
        "temp": { "type": "boolean" }
      },
      "required": ["type"],
      "additionalProperties": false
    },
    {
      "type": "object",
      "properties": {
        "type": { "const": "LocalIp" },
        "showIpv6": { "type": "boolean" },
        "namePrefix": { "type": "string" }
      },
      "required": ["type"],
      "additionalProperties": false
    }
  ]
}
```

In most validators, you won't get very good error messages with this pattern in some cases, but editors/IDEs tend to handle this kind of thing pretty well, so I think this will be a good approach for you.

**Reply by CarterLi** [reply]
Completely the same as what I did 😄

**Comment by CarterLi** [comment]
Another question

Can I make enums / constants case insensitive?

**Reply by jdesrosiers** [reply]
Unfortunately, there isn't a good solution for that. I think the best you can do is to use regular expressions something like this.

```json
{
  "$defs": {
    "my-enum": {
      "anyOf": [
        { "type": "string", "pattern": "^[fF][oO][oO]$" },
        { "type": "string", "pattern": "^[bB][aA][rR]$" }
      ]
    }
  }
}
```
